### PR TITLE
Add MonadReader, instances for Function1 and Kleisli

### PR DESCRIPTION
**Pull request by adelbertc**

### CODE REVIEW

File: core/src/main/scala/cats/MonadReader.scala [hunk 1]
```diff
@@ -0,0 +1,14 @@
+package cats
+
+/** A monad that has the ability to read from an environment. */
+trait MonadReader[F[_, _], R] extends Monad[F[R, ?]] {
+  /** Get the environment */
+  def ask: F[R, R]
+
+  /** Modify the environment */
+  def local[A](f: R => R)(fa: F[R, A]): F[R, A]
```
Comment (ceedubs) [inline core/src/main/scala/cats/MonadReader.scala]: Should we make this `def local[A](fa: F[R, A])(f: R => R): F[R, A]` to line up with the definitions for `map`, `flatMap`, `traverse`, etc in other type classes?

**Comment by ceedubs** [comment]
Other than my one stylistic question, this looks good to me.

File: core/src/main/scala/cats/MonadReader.scala (hunk 1 above)
Comment (adelbertc) [inline core/src/main/scala/cats/MonadReader.scala]: I am OK with either, one thing I just noticed is that the direction may or may not look nicer due to the "flow" of the operation. For instance,

```scala
[M[_], A]def flatMap[B](ma: M[A])(f: A => M[B]): M[B]
```
We run the effects of `ma` first, and then apply `f` to it, and then run the effects of `M[B]`, so the flow is left to right. similarly for `map` and `traverse`.

With local:

```scala
[F[_, _], R]def local[A](f: R => R)(fa: F[R, A]): F[R, A]
```

We are first modifying the environment with `f` and then running `fa` against that, as opposed to running `fa` and then modifying the environment after the fact.

Comment (ceedubs) [inline core/src/main/scala/cats/MonadReader.scala]: Ah your point about the flow is a good one. I didn't think about that.

One reason I like about putting the function in the last parameter list is that I often am using a lambda that I format as multiple lines:

```scala
(...).local(fra)(r =>
  doStuffWithR(r))
```

Which looks slightly nicer to me than:

```scala
(...).local(r =>
  doStuffWithR(r)
)(fra)
```

So between the convention on other methods, and this formatting, I'd be a bit inclined to have the function in a second parameter list; but it's not a big deal to me. Whichever people prefer.

Comment (adelbertc) [inline core/src/main/scala/cats/MonadReader.scala]: Fair enough, I often find myself really liking the function standalone in the last parameter list for the exact reason you describe above.

I'll wait for others to chime in with their thoughts, if folks like it as the 2nd/last parameter I'll happily make the change :-)

Comment (stew) [inline core/src/main/scala/cats/MonadReader.scala]: do we worry about confusion of the method name local with the Kleisli#local from scalaz (which is contramap)?

Comment (stew) [inline core/src/main/scala/cats/MonadReader.scala]: If I have to choose a side in the arg list ordering, I'd probably prefer Chang's.

Partially applying a function to get back a Kleisli => Kleisli seems more natural to me than partially applying a Kleisli and getting back a (R => R) => Kleisli

Comment (ceedubs) [inline core/src/main/scala/cats/MonadReader.scala]: Okay, I have no problem with that.

Comment (adelbertc) [inline core/src/main/scala/cats/MonadReader.scala]: @stew Yeah I had wondered that we have `Kleisli#local: Kleisli[M, A, B] => (AA => A) => Kleisli[M, AA, B]` (which I find useful in "expanding" the config when I use it in a more global scope), but also `MonadReader#local: M[R, A] => (R => R) => M[R, A]`. Perhaps we can rename the `Kleisli#local` to something else? ... `expand` ?

**Comment by ceedubs** [comment]
:+1: I think we can always sort out any `local` naming changes later if it feels wrong.

**Comment by non** [comment]
I agree with @ceedubs. :+1:

**Pull request merged**